bootstrap/src/main/java/jenkins/bootstrap/Plugin.java

```diff
@@ -0,0 +1,188 @@
+                    continue;
+                // TODO: version check
+                jars.add(jar.toURI().toURL());
+            }
+        }
+    }
+
+    /**
+     * Gets the directory name to explode the jpi file into.
+     */
+    private File getDestDir() {
+        String name = archive.getName();
+        int idx = name.lastIndexOf('.');
+        if (idx<0)  throw new IllegalArgumentException("No extension in file name: "+archive);
+        return new File(archive.getParentFile(), name.substring(0,idx));
+    }
+
+    /**
+     * Returns true if this plugin overrides core components.
+     */
+    private boolean isCoreComponentOverride() throws IOException {
+        Manifest mf = loadManifest();
+        return "true".equals(mf.getMainAttributes().getValue("Core-Component"));
+    }
+
+    public boolean isDisabled() {
+        File disableFile = new File(archive.getPath() + ".disabled");
+        return disableFile.exists();
+    }
+
+    /**
+     * Loads jar file manifest from the archive.
+     */
+    private Manifest loadManifest() throws IOException {
+        try (JarFile jar = new JarFile(archive)) {
+            ZipEntry e = jar.getEntry("META-INF/MANIFEST.MF");
+            try (InputStream is = jar.getInputStream(e)) {
+                return new Manifest(is);
+            }
+        }
+    }
+
+    /**
+     * Unpacks {@link #archive} into {@link #destDir} with uptodate check.
+     *
+     * <p>
+     * Up-to-date check needs to be consistent with {@coe ClassicPluginStrategy} in core.
+     *
+     * <p>
+     * This simplified version does not support automatic building of classes.jar from WEB-INF/classes.
+     * Pluggable core components are not supposed to have any classes, just WEB-INF/lib files.
+     */
+    private void explode() throws IOException {
+        destDir.mkdirs();
+
+        // timestamp check
+        File explodeTime = new File(destDir,".timestamp2");
+        if(explodeTime.exists() && explodeTime.lastModified()==archive.lastModified())
+            return; // no need to expand
+
+        Project p = new Project();
+
+        if (destDir.exists()) {
+            // delete the contents so that old files won't interfere with new files
+            Delete d = new Delete();
+            d.setProject(p);
+            if (destDir.isDirectory())
+                d.setDir(destDir);
+            else
+                d.setFile(destDir);
+            d.execute();
+        }
+
+        // expand the contents
+        Expand e = new Expand();
+        e.setProject(p);
+        e.setTaskType("unzip");
+        e.setSrc(archive);
+        e.setDest(destDir);
+        e.execute();
+
+        if (new File(destDir,"WEB-INF/classes").exists()) {
+            throw new IOException("Core override plugin cannot have WEB-INF/classes: "+destDir);
+        }
+
+        // update timestamp file
+        new FileOutputStream(explodeTime).close();
+        explodeTime.setLastModified(archive.lastModified());
```

Comment:
This seems risky, there are other things that could be touching the file. Might make more sense to just store the timestamp as a long value in the file to guard against both permission issues with some file systems (i.e. you may not be able to set the lastModified) and guard against other processes using lastModified for their own (evil) purposes